utilities/ggd/main.go

```diff
@@ -0,0 +1,360 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package main
+
+import (
+	"fmt"
+	"go/build"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/m3db/build-tools/utilities/ggd/lib"
+
+	"github.com/pborman/getopt"
+	"github.com/tmc/dot"
+)
+
+const usageNotes = `ggd: command line tool to find packages affected by git changes. Examples:
+# Assuming CWD is in a git repository directory present in the GOPATH.
+
+# (1) List all the golang packages affected by changes between master and HEAD
+ggd
+
+# (2) List all the golang packages affected by changes between branchA and HEAD
+ggd branchA
+
+# (3) List all the golang packages affected by between changes between branchA and branchB
+ggd branchA..branchB
+
+# (4) Run tests for all the golang packages affected by between master, and head
+go test $(ggd)
+
+# (5) List all the golang packages affected by between master and head, and build tags 'integration'
+ggd -t integration
+
+# (6) The same as (5), but include debug output (sent to STDERR)
+ggd -t integration -d
+
+# (7) The same as (6), but include debug output (sent to STDERR), and
+# save the generated DAG in changes.png for visualization
+ggd -t integration -d -o change.dot
+dot -Tpng change.png change.dot
+`
+
+var (
+	basePkg          = getopt.StringLong("base-package", 'p', "", "repository package (can leave blank typically)")
+	debugMode        = getopt.BoolLong("debug", 'd', "debug mode - useful for understanding output")
+	buildTags        = getopt.ListLong("build-tags", 't', "comma separated golang build tags")
+	catchallPatterns = getopt.ListLong("catchall-patterns", 'c', "comma separated file/dir names, such that if any of these are changed, all tests are run")
+	filterPatterns   = getopt.ListLong("filter-patterns", 'f', "comma separated dir names to be filtered out during change calculation")
+	dotOutputFile    = getopt.StringLong("output-file", 'o', "", "optionally, visualize changes in a dot file (only in debug mode)")
+)
+
+var (
+	defaultCatchallPatterns = []string{
+		"/testdata/", // go list ./... is unable to gauge impact of changes to this directory
+	}
+	defaultFilterPatterns = []string{"/vendor/"}
```

Comment:
Should the default for this be empty?

Reply:
No. Going to stick with this as the default because it's typically what people want when using the tool by hand. I prefer to optimise for that over what the CI invocation for this will be. 

`go test` does this too so it's not unexpected behaviour.

Reply:
Hmm, but if working locally and I update a vendored dependency wouldn't it be preferable to test the packages which depend on the changed vendor package?

Reply:
Sure. But in that case, you've updated `glide.yaml/glide.lock` (added to catchall patterns) and we'd run all tests

Reply:
or, they can alternatively run with `ggd -f=''` and get the same behaviour (w/o using glide changes as catch all)